source/extensions/filters/http/oauth2/oauth_client.cc

```diff
@@ -21,21 +22,38 @@ namespace HttpFilters {
 namespace Oauth2 {
 
 namespace {
-constexpr const char* GetAccessTokenBodyFormatString =
+constexpr const char* GetAccessTokenBodyFormatString_01 =
     "grant_type=authorization_code&code={0}&client_id={1}&client_secret={2}&redirect_uri={3}";
 
+constexpr const char* GetAccessTokenBodyFormatString_02 =
+    "grant_type=authorization_code&code={0}&redirect_uri={1}";
+
 } // namespace
 
 void OAuth2ClientImpl::asyncGetAccessToken(const std::string& auth_code,
                                            const std::string& client_id, const std::string& secret,
-                                           const std::string& cb_url) {
+                                           const std::string& cb_url, AuthType auth_type) {
   const auto encoded_client_id = Http::Utility::PercentEncoding::encode(client_id, ":/=&?");
   const auto encoded_secret = Http::Utility::PercentEncoding::encode(secret, ":/=&?");
   const auto encoded_cb_url = Http::Utility::PercentEncoding::encode(cb_url, ":/=&?");
 
   Http::RequestMessagePtr request = createPostRequest();
-  const std::string body = fmt::format(GetAccessTokenBodyFormatString, auth_code, encoded_client_id,
-                                       encoded_secret, encoded_cb_url);
+  std::string body;
+  switch (auth_type) {
+  case AuthType::UrlEncodedBody:
+    body = fmt::format(GetAccessTokenBodyFormatString_01, auth_code, encoded_client_id,
+                       encoded_secret, encoded_cb_url);
+    break;
+  case AuthType::BasicAuth:
+    const auto basic_auth_token = absl::StrCat(client_id, ":", secret);
+    const auto encoded_token = Base64::encode(basic_auth_token.data(), basic_auth_token.size());
+    const auto basic_auth_header_value = absl::StrCat("Basic ", encoded_token);
```

Comment:
Are we able to pre-encode this early on the way we do for the other parameters? That would save us doing this on every request